File: src/windows/magma/app/Generic.c
#include "./Generic.h"
class Generic implements Type {
	mut base: &[I8];
	mut args: List<&[I8]>;
	constructor (mut base: &[I8], mut args: List<&[I8]>) {
		this.base = base;
		this.args = args;
	}
	mut static generateValueStrings(values: List<&[I8]>): &[I8] {
		return Main.generateAll(values, Generic.mergeValues);
	}
	mut static mergeValues(cache: &[I8], element: &[I8]): &[I8] {
		if (Strings.isEmpty(cache)) {
			return cache + element;
		}
		return cache + ", " + element;
	}
	mut generate(): &[I8] {
		return this.base + "<" + Generic.generateValueStrings(this.args) + ">";
	}
	mut isFunctional(): Bool {
		return false;
	}
	mut isVar(): Bool {
		return false;
	}
	mut generateBeforeName(): &[I8] {
		return "";
	}
}

File: src/java/magma/app/Generic.java
package magma.app;

import jvm.api.text.Strings;
import magma.api.Type;
import magma.api.collect.list.List;

record Generic(String base, List<String> args) implements Type {
    private static String generateValueStrings(final List<String> values) {
        return Main.generateAll(values, Generic::mergeValues);
    }

    private static String mergeValues(final String cache, final String element) {
        if (Strings.isEmpty(cache)) {
            return cache + element;
        }
        return cache + ", " + element;
    }

    @Override
    public String generate() {
        return this.base + "<" + Generic.generateValueStrings(this.args) + ">";
    }

    @Override
    public boolean isFunctional() {
        return false;
    }

    @Override
    public boolean isVar() {
        return false;
    }

    @Override
    public String generateBeforeName() {
        return "";
    }
}
